### PR TITLE
Add child_storage and getKeys methods

### PR DESCRIPTION
**Pull request by Stefie**
Solves https://github.com/polkadot-js/api/issues/930

(Not testable in apps because of https://github.com/polkadot-js/apps/issues/1281)

### CODE REVIEW

File: packages/types/src/scripts/METHODS_STORAGE_SUBSTRATE.md [hunk 1]
```diff
@@ -14,6 +14,9 @@ _These are keys that are always available to the runtime implementation_
 ▸ **changesTrieConfig**(): `u32`
 - **summary**: Changes trie configuration is stored under this key.
 
+▸ **childStorageKeyPrefix**(): `u32`
```
Comment (jacogr) [inline packages/types/src/scripts/METHODS_STORAGE_SUBSTRATE.md]: Remove, auto-generated, done by CI.

Comment (Stefie) [reply packages/types/src/scripts/METHODS_STORAGE_SUBSTRATE.md]: This is not auto generated, it generates `docs/METHODS_STORAGE.md ` according to your comment here https://github.com/polkadot-js/api/pull/1030#discussion_r294729850 .

Comment (jacogr) [reply packages/types/src/scripts/METHODS_STORAGE_SUBSTRATE.md]: Indeed. I get very jittery with all manual edits to `*.md` files, so misread where this one was from after a barrage elsewhere :)

File: packages/rpc-core/test/e2e/state.test.ts [hunk 2]
```diff
@@ -2,10 +2,20 @@
...
-import { Balance, Bytes, Metadata, Moment } from '@polkadot/types';
+import fs from 'fs';
+import path from 'path';
+
+import { Balance, Bytes, Hash, Metadata, Moment, StorageData, StorageKey } from '@polkadot/types';
+import { Abi } from '@polkadot/api-contract';
 import storage from '@polkadot/storage/static';
 import WsProvider from '@polkadot/rpc-provider/ws';
 
+import ApiPromise from '@polkadot/api/promise/Api';
```
Comment (jacogr) [inline packages/rpc-core/test/e2e/state.test.ts]: (nit) Alphabetical ordering (yes, these are tests, but in this case we are starting to do a lot of imports)

File: packages/rpc-core/test/e2e/state.test.ts [hunk 3]
```diff
@@ -2,10 +2,20 @@
...
+import { SubmittableResult } from '@polkadot/api';
+import { KeyringPair } from '@polkadot/keyring/types';
+import testingPairs from '@polkadot/keyring/testingPairs';
+import flipperAbi from '../../../api-contract/test/contracts/flipper.json';
```
Comment (jacogr) [inline packages/rpc-core/test/e2e/state.test.ts]: (nit) Relative imports with a blank line from the package imports.

File: packages/rpc-core/test/e2e/state.test.ts [hunk 4]
```diff
@@ -2,10 +2,20 @@
...
+
```
Comment (jacogr) [inline packages/rpc-core/test/e2e/state.test.ts]: (nit) Remove this extra space, goes with above one (which should have a blank line preceding it).

File: packages/rpc-core/test/e2e/state.test.ts [hunk 5]
```diff
@@ -46,37 +41,146 @@ describe.skip('e2e state', () => {
...
 
-  it('retrieves balances', () => {
+  it('getKeys(): retrieves storage keys for ":code"', () => {
     return api.state
-      .getStorage([
-        storage.balances.freeBalance, ALICE
-      ])
-      .then((balance: Balance) => {
-        console.error(balance);
+      .getKeys('0x3a636f6465')
```
Comment (jacogr) [inline packages/rpc-core/test/e2e/state.test.ts]: Even though this is a test, may make sense to put this below the ADDR above (as `cost CODE = '0x3a636f6465'; // :code`)

File: packages/rpc-core/test/e2e/state.test.ts [hunk 6]
```diff
@@ -46,37 +41,146 @@ describe.skip('e2e state', () => {
...
+          }
+          done();
+        });
+
+      return putCode.then(() => {
+        return apiPromise.tx.contract
+          .create(12345, 50000, codeHash, abi.deploy())
+          .signAndSend(keyring.bob);
+      }).catch(e => {
+        console.log(e);
```
Comment (jacogr) [inline packages/rpc-core/test/e2e/state.test.ts]: The test should already log it? (Can remove the catch and rethrow here)

File: packages/rpc-core/test/e2e/state.test.ts [hunk 7]
```diff
@@ -46,37 +41,146 @@ describe.skip('e2e state', () => {
...
+    it('getChildStorage(): retrieves the default value of the flipper smart contract', async () => {
+      const storageKeys = await api.state.getKeys('0x3a6368696c645f73746f726167653a');
+      const childStorageKeys = await api.state.getChildKeys(storageKeys[0], '0x');
+
+      return api.state
+        .getChildStorage(storageKeys[0], childStorageKeys[0])
+        .then((storage: StorageData) => {
+          expect(storage.toString()).toBe('0x00');
+        });
+
```
Comment (jacogr) [inline packages/rpc-core/test/e2e/state.test.ts]: (nit) Remove blank line

File: packages/rpc-core/test/e2e/state.test.ts [hunk 8]
```diff
@@ -46,37 +41,146 @@ describe.skip('e2e state', () => {
...
+    it('getChildStorageHash(): retrieves the Hash of the flipper smart contract', async () => {
+      const storageKeys = await api.state.getKeys('0x3a6368696c645f73746f726167653a');
+      const childStorageKeys = await api.state.getChildKeys(storageKeys[0], '0x');
+
+      return api.state
+        .getChildStorageHash(storageKeys[0], childStorageKeys[0])
+        .then((storage: StorageData) => {
+          expect(storage.toString()).toBe('0x03170a2e7597b7b7e3d84c05391d139a62b157e78786d8c082f29dcf4c111314');
+        });
+
```
Comment (jacogr) [inline packages/rpc-core/test/e2e/state.test.ts]: (nit) Remove blank line

File: packages/rpc-core/test/e2e/state.test.ts [hunk 9]
```diff
@@ -46,37 +41,146 @@ describe.skip('e2e state', () => {
...
+    it('getChildStorageSize(): retrieves the size of the flipper smart contract', async () => {
+      const storageKeys = await api.state.getKeys('0x3a6368696c645f73746f726167653a');
+      const childStorageKeys = await api.state.getChildKeys(storageKeys[0], '0x');
+
+      return api.state
+        .getChildStorageSize(storageKeys[0], childStorageKeys[0])
+        .then((storage: StorageData) => {
+          expect(storage.toString()).toBe('1');
+        });
+
```
Comment (jacogr) [inline packages/rpc-core/test/e2e/state.test.ts]: (nit) remove blank line

File: packages/rpc-core/test/e2e/state.test.ts [hunk 10]
```diff
@@ -46,37 +41,146 @@ describe.skip('e2e state', () => {
...
+    });
+
```
Comment (jacogr) [inline packages/rpc-core/test/e2e/state.test.ts]: (nit) remove blank line

File: packages/type-jsonrpc/src/state.ts [hunk 11]
```diff
@@ -19,6 +19,57 @@ const call: RpcMethodOpt = {
   type: 'Bytes'
 };
 
+const getChildKeys: RpcMethodOpt = {
```
Comment (jacogr) [inline packages/type-jsonrpc/src/state.ts]: Can we align the ordering here to match what is in substrate? It looks very arbitrary, but it helps an immense amount when going through and looking at discrepancies and missing methods - there are a lot. 

(For the exports below, right at the end, we do it alphabetical - so it is sane)

File: packages/rpc-core/test/e2e/state.test.ts (hunk 5 above)
Comment (Stefie) [reply packages/rpc-core/test/e2e/state.test.ts]: done.

File: packages/rpc-core/test/e2e/state.test.ts [hunk 12]
```diff
@@ -19,7 +19,7 @@ import flipperAbi from '../../../api-contract/test/contracts/flipper.json';
 
 const ALICE = '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY';
 const CODE = '0x3a636f6465'; // :code
-const CHILD_STORAGE = '0x3a636f6465'; // :child_storage:
+const CHILD_STORAGE = '0x3a6368696c645f73746f726167653a'; // :child_storage:
```
Comment (jacogr) [inline packages/rpc-core/test/e2e/state.test.ts]: +1

**Pull request merged**